common/lockless_queue.c

```diff
@@ -0,0 +1,143 @@
+#ifdef _MSC_VER
+    oe_lockless_queue_node* p_actual = NULL;
+#endif /* _MSC_VER */
+    do
+    {
+#ifdef _MSC_VER
+        p_expected = p_actual;
+        p_node->p_link = p_expected;
+        p_actual = (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+            &(p_queue->p_tail), p_node, p_expected);
+    } while (p_actual != p_expected);
+#elif defined __GNUC__
+        p_node->p_link = p_expected;
+    } while (!__atomic_compare_exchange_n(
+        &(p_queue->p_tail),
+        &p_expected,
+        p_node,
+        1,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_ACQUIRE));
+#endif /* __GNUC__ */
+} /* oe_lockless_queue_push */
+
+oe_lockless_queue_node* oe_lockless_queue_pop_front(oe_lockless_queue* p_queue)
+{
+    /* try to take a node from the head */
+    oe_lockless_queue_node* popped_node = NULL;
+#ifdef _MSC_VER
+    popped_node = (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+        &(p_queue->p_head), NULL, NULL);
+#elif defined __GNUC__
+    popped_node = __atomic_load_n(&(p_queue->p_head), __ATOMIC_ACQUIRE);
+#endif /* __GNUC__ */
+
+    if (NULL != popped_node)
+    {
+        /* there was a node at the head
+         * pop the node from the head and replace it with the node that it
+         *     points to
+         * remove the reference from the popped node to the next node */
+        oe_lockless_queue_node* next_node = popped_node->p_link;
+        popped_node->p_link = NULL;
+#ifdef _MSC_VER
+        _InterlockedExchangePointer(&(p_queue->p_head), next_node);
+#elif defined __GNUC__
+        __atomic_store_n(&(p_queue->p_head), next_node, __ATOMIC_RELEASE);
+#endif /* __GNUC__ */
+    }
+    else
+    {
+        /* there wasn't a node at the head
+         * so refill the head with the nodes from the tail */
+#ifdef _MSC_VER
+        popped_node =
+            (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+                &(p_queue->p_tail), NULL, NULL);
+#elif defined __GNUC__
+        popped_node = __atomic_load_n(&(p_queue->p_tail), __ATOMIC_RELAXED);
+#endif /* __GNUC__ */
+
+        if (NULL != popped_node)
+        {
+            /* take all of the nodes off of the tail */
+#ifdef _MSC_VER
+            oe_lockless_queue_node* p_actual = NULL;
+            while (popped_node !=
+                   (p_actual = (oe_lockless_queue_node*)
+                        _InterlockedCompareExchangePointer(
+                            &(p_queue->p_tail), NULL, popped_node)))
+            {
+                popped_node = p_actual;
+            }
+#elif defined __GNUC__
+            while (!__atomic_compare_exchange_n(
+                &(p_queue->p_tail),
+                &popped_node,
+                NULL,
+                1,
+                __ATOMIC_ACQ_REL,
+                __ATOMIC_ACQUIRE))
+            {
+                continue;
+            }
+#endif /* __GNUC__ */
+
+            /* reverse the nodes from the tail */
+            oe_lockless_queue_node* prev_node = NULL;
```

Comment:
I don't understand why we need to pop all the nodes and re-insert them.
What happens to the order when a new node is added while we are in the process of re-adding all these nodes here?

Reply:
This is because of the two stack implementation.  https://github.com/microsoft/openenclave/pull/1838#discussion_r288771644  Two stacks are used because it is not possible to update more than one pointer with a single atomic operation.  It is also not possible to implement a double-linked list atomically.  However, this implementation is safe for pushing a new node while concurrently popping a node even in the case where all the nodes are being moved from the tail to the head.

Note:  I have implemented lockless FIFO queues before as a single-linked list.  While that makes it much easier to follow the logic, it is much less efficient because each pop needs to iterate through each node in the list (using full atomic synchronization for each node->p_next) to get to the head.  That implementation has O (n x n) complexity while the two stack implementation has O (n) complexity.